msal/src/test/java/com/microsoft/identity/client/robolectric/tests/RoboTestUtils.java

```diff
@@ -0,0 +1,113 @@
+     *
+     * @param cacheKey The cache key to inspect.
+     * @return The CredentialType or null if a proper type cannot be resolved.
+     */
+    @Nullable
+    private static CredentialType getCredentialTypeForCredentialCacheKey(@NonNull final String cacheKey) {
+        if (StringExtensions.isNullOrBlank(cacheKey)) {
+            throw new IllegalArgumentException("Param [cacheKey] cannot be null.");
+        }
+
+        final Set<String> credentialTypesLowerCase = new HashSet<>();
+
+        for (final String credentialTypeStr : CredentialType.valueSet()) {
+            credentialTypesLowerCase.add(credentialTypeStr.toLowerCase(Locale.US));
+        }
+
+        CredentialType type = null;
+        for (final String credentialTypeStr : credentialTypesLowerCase) {
+            if (cacheKey.contains(CACHE_VALUE_SEPARATOR + credentialTypeStr + CACHE_VALUE_SEPARATOR)) {
+                if (credentialTypeStr.equalsIgnoreCase(CredentialType.AccessToken.name())) {
+                    type = CredentialType.AccessToken;
+                    break;
+                } else if (credentialTypeStr.equalsIgnoreCase(CredentialType.RefreshToken.name())) {
+                    type = CredentialType.RefreshToken;
+                    break;
+                } else if (credentialTypeStr.equalsIgnoreCase(CredentialType.IdToken.name())) {
+                    type = CredentialType.IdToken;
+                    break;
+                } else if (credentialTypeStr.equalsIgnoreCase(CredentialType.V1IdToken.name())) {
+                    type = CredentialType.V1IdToken;
+                    break;
+                }
+            }
+        }
+
+        return type;
+    }
+
+    private static boolean isAccessToken(@NonNull final String cacheKey) {
+        boolean isAccessToken = CredentialType.AccessToken == getCredentialTypeForCredentialCacheKey(cacheKey);
+        return isAccessToken;
+    }
+
+    private static SharedPreferences getSharedPreferences() {
+        final Context context = ApplicationProvider.getApplicationContext();
+        SharedPreferences sharedPreferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE);
+        return sharedPreferences;
+    }
+
+    static void clearCache() {
+        SharedPreferences sharedPreferences = getSharedPreferences();
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.clear();
+        editor.commit();
+    }
+
+    static void removeAccessTokenFromCache() {
+        SharedPreferences sharedPreferences = getSharedPreferences();
+        final Map<String, ?> cacheValues = sharedPreferences.getAll();
+        final String keyToRemove = getCacheKeyForAccessToken(cacheValues);
+        if (keyToRemove != null) {
+            SharedPreferences.Editor editor = sharedPreferences.edit();
+            editor.remove(keyToRemove);
+            editor.commit();
+        }
+    }
+
+    static void flushScheduler() {
```

Comment:
Throwing out an alternative here.

Would it be cleaner to use countDownLatch instead (and rename it to waitForResult() or something)
- AcquireTokenBase calls waitForResult(), which will block the thread.
- in every exit point, we invoke returnResultPass() or returnResultFail() which essentially will unblock the thread. 

In this case we just need to call waitToResult() once per test case, at the very last line.

Reply:
I can rename the method for sure, but countDownLatch does not work well with Robolectric. In my experience, I found it much easier to work with flushing the scheduler than using the countDownLatch with Robolectric.

Reply:
Also if we use something like returnResultPass() or returnResultFail() then we would also have multiple calls to those then for every exit point, and since there would be multiple exit points that would also add up to a lot of method calls. I'll prefer sticking to the current approach.